### PR TITLE
fix #9352: transpose by range (for 2.1)

### LINKED ISSUE #9352: [MU4 Issue] Properly determine 'midpoint' of noteheads for origin of ties
This is less critical for 'normal' noteheads, which are round, though as they are somewhat oblique the centre at the top is optically a little further right than the centre at the bottom.

However, it is important for slashes, for example:

slashes

The cutout metadata may be able to help with this; failing that, perhaps slashes might be special-cased.

**Should fix:**
https://musescore.org/en/node/325037

### PR DESCRIPTION
**Pull request by MarcSabatella**
This is my initial cut at incorporating my previous code from master into 2.1, to implement the ability for instrument change to affect transposition.

Right now the code works for me as well as it does on master, but there are some issues I want to investigate and hopefully see resolved.  See in particular:

https://musescore.org/en/node/78591
https://musescore.org/en/node/62416

See in particular my comment in the latter: https://musescore.org/en/node/62416#comment-664481.  Not only do staff properties changes not propagate, neither do instrument change commands.  So if you add an instrument change text to the score and then do a change instrument on it, it works in the score, and the text appears in the part, but it has no effect on playback (or transposition).  This affects 2.0.3 and probably should be filed separately, but I'm not sure yet if it's really just another symptom of the same issue or not.

Right now I am still struggling with build issues - I can build but not debug on my branch, apparently my build is not getting any debug info despite setting CMAKE_BUILD_TYPE=Debug.  So I have some work to do before I can really continue with this.  But if anyone else wants to try out this code and investigatye those issues at all, please be my guest.

### CODE REVIEW

**Comment by MarcSabatella** [comment]
Finally have a build going, and debugger mostly working.

I added code to fix one of the outstanding issues with instrument changes: the fact that the "change instrument" command does not affect linked parts.  My code works for now, but I assume it will need to be dealt with differently if/when we fix the undo issue.

**Comment by Jojo-Schmitz** [comment]
mtest gets an exception (crash?) in tst_instrumentchange

**Comment by lasconic** [comment]
copy.mscx is too new and cannot be read with MuseScore 2. Changing the version from 2.07 to 2.06 should make it pass.

**Comment by MarcSabatella** [comment]
I made that change.  I will want to further update the tests once I'm sure about the approach I am taking regarding compatibility for existing scores.  See https://musescore.org/en/node/177886 for a discussion I started in an effort to collect input.  Not much feedback, which I guess I will take as a good sign - that maybe people are "saying" they just want it fixed and don't much case how I handle existing scores.

Here is what I am currently doing:

1) In InstrumentChange::read(), I use mscoreVersion() to check for 2.0-based scores.  For these I explicitly ignore the transposition in the InstrumentChange and reset it to whatever was currently in effect.  This mimics the behavior of 2.0.x.  Thus, changing from Bb clarinet to Eb saxophone will change sound but not transposition, just as it always has.  The tpc corruption that was introduced in 2.0.x if you input notes after adding the instrument change will be automatically corrected during the note read.

2) In Note::read() I check for tpc corruption - places where one or more of pitch, tpc1, and tpc2 are not what they should be according to the transposition currently in effect.  I am currently resolving conflicts by assuming that whatever is actually shown in the score is what you want.  So if concert pitch is turned *on*, it is the sounding pitch we want to preserve, so I adjust the written pitch accordingly.  Whereas if concert pitch is turned *off*, it is the written pitch we want to preserve, so I adjust the sounding pitch accordingly.

None of this is set in stone: I could easily tweak it to do pretty much whatever people say they want.  I am strongly considering *not* doing what I described in 1) at all.  Doing that basically means existing scores don't get to take advantage of the new behavior, unless the user goes in and re-does the "change instrument".  But it does mean that if the user had employed some sort of hack/workaround in an effort to get what he wanted, that should continue to work.

**Comment by lasconic** [comment]
*But it does mean that if the user had employed some sort of hack/workaround in an effort to get what he wanted, that should continue to work.*
What sort of hack/workaround are known ?

**Comment by MarcSabatella** [comment]
Well, that's part of why I wanted feedback from the forum - to see what people *are* doing.

What I personally have done is have two staves, one set to invisible but with correct playback, the other muted but with correct appearance.  I think I've seen others mentioning using the "Fix to Line" property to alter the appearance of notes, which seems a bit drastic.

**Comment by lasconic** [comment]
So in your case, the appearance will not be good anymore and you will have to review your scores but in the fix to line case, it will work. Correct?

**Comment by MarcSabatella** [comment]
If I do what is in #1, then my existing scores should behave exactly the same, except to the extent I might have tpc corruption that will now be fixed.  If I want to take advantage of the transposition feature for real, I would expect to need to redo the instrument change and then assess from there.

If I *don't* do what is in #1, then assuming I had saved my score with concert pitch *off* (which would be the norm for any score I had completed), then the staff that I had designed to *appear* correctly but playback incorrectly will still appear correctly, but it will now *also* playback correctly (because I am honoring the displayed pitches and changing playback pitches).  The staff I had set up to play back correctly but appear incorrectly will now do *both* incorrectly, but I no longer need that staff so I can just delete it.  Or I could just manually transpose it and it will now also do both correctly.

If I save a score with concert pitch *on*, that's generally because I am still working on it.  In these cases, I generally will have entered correct concert pitches onto all staves; I'm not yet to the point of employing workarounds.  So in these cases I would want the concert pitch to be what is preserved - which is exactly what my current algorithm does.

So for me personally, what I did is pretty much exactly what I'd want, except I personally would just as soon *not* do what is in #1.  I'm only including it for the benefit of people who might want to see as little change as possible - assuming such people exist.  Again, that's why I was hoping for feedback.

**Comment by MarcSabatella** [comment]
Still not there 100%, but I think I found a possible key to the undo bug that has been dogging me.

I had already realized the reason undo wasn't working at all for the change instrument command is that we were setting the new instrument on the instrument change before we create the undo object, here:

https://github.com/musescore/MuseScore/blob/2.1/mscore/propertymenu.cpp#L556

But just today I noticed that when creating the instrument change element in the first place, we don't call setInstrument() for the instrument change element itself, we do it for the *part* the element is attached to, here:

https://github.com/musescore/MuseScore/blob/2.1/libmscore/score.cpp#L1437

Simply making the change instrument command do this was suddenly gets things working much better, but still not perfect.  I get funky incorrect key changes on if I then do a series of undo / redo operations, and a crash if I keep pushing this at the beginning of a score where there is only the "implied" C major key signature.  But I think I'm making progress.  If anyone wants to try further, be my guest...

**Comment by MarcSabatella** [comment]
Update:

I am now realizing that transpositionChanged() probably shouldn't be called from within an undo action (eg, from within ChangeInstrument::flip()), because transpositionChanged() will itself kick off undo-able actions, resulting in things basically being undone twice.  I didn't notice this when originally implementing this for 3.0 because the change instrument command was not being undone properly, but now that I have the basic change instrument undo working, I realize I'm actually getting these double undos.

So now I have now moved the transpositionChanged() call out of the ChangeInstrument::flip() code and into the property menu handler, which also mimics how transpositionChanged is used in the Staff Properties dialog.  I need to do some more testing and code cleanup, but I am now very optimistic that I will have this working with proper undo within the next couple of days.

If I can get that much ready, I'd consider this worth merging.  There is still the issue of how instrument changes affect parts (https://musescore.org/en/node/62416), but it's basically the same issue that already exists regarding changing the instrument in staff properties.  I'd love it is someone could look into that, unfortunately I am probably not going to have time before 2.1 as next week I take off for most of the rest of the month.

**Comment by MarcSabatella** [comment]
I believe I have everything working as well as can be with respect to instrument change elements.  Undo of the change instrument command now works properly, and the command now affects linked parts properly as well.  Also, copy/paste on the element works properly; it also could have had issues previously.  All of these were kind of prerequisites to getting the transposition code working well.

It's still the case that the basic change instrument command within *staff properties* (totally separate from the instrument change element I am dealing with here) does not affect linked parts properly.  See https://musescore.org/en/node/62416.  My changes here improve the situation somewhat - the key signature is now transposed along with the notes, and this in turn also corrects the octave errors in the transposition of the notes.  However, the transposition is still not set for the staff in the parts, meaning we still basically have tpc corruption.  With my changes in place - specifically to Score::transpositionChanged() -  it should now be pretty easy to deal with that, but I'll leave that for another day.

So, I do think this code is good and we can consider merging it for 2.1.  I realize it may be a bigger change than we are comfortable with for 2.1 at this point, but I'm putting it out there anyhow.  If nothing else, we will want to re-absorb this code into  3.0, as my undo and parts fixes will be needed there.

**Comment by MarcSabatella** [comment]
(rebased, not that this was technically necessary)

**Comment by lasconic** [comment]
If you have 5 minutes in the coming days, could you add a paragraph to https://musescore.org/en/developers-handbook/release-notes/release-notes-musescore-2.1#known-incompatibilities to describe how files created with 2.1 would open differently in 2.0.X?

**Comment by MarcSabatella** [comment]
No problem, will get to that soon.  I haven't thought as much about going that direction, but I think the story will actually be pretty good.

**Comment by MarcSabatella** [comment]
Done.  The gist of it is, 2.0 scores (which are guaranteed to be incorrect for display, playback, or both) will continue to be incorrect if loaded into 2.1, although you can then use 2.1 to fix them and they will work correctly from then on within 2.1.  But 2.1 scores (which will display and playback correctly in 2.1) will not continue to do work correctly if loaded into 2.0.  Playback will generally be correct, but display will be incorrect much as it would be if you had created the score in 2.0 in the first place and had prioritized getting proper playback over getting proper display.

File: libmscore/instrchange.cpp [hunk 1]
```diff
@@ -87,6 +88,19 @@ void InstrumentChange::read(XmlReader& e)
             else if (!Text::readProperties(e))
                   e.unknown();
             }
+      QString sv = score()->mscoreVersion();
+      sv.truncate(3);
+      if (sv == "2.0") {
```
Comment (lasconic) [inline libmscore/instrchange.cpp]: I really don't like this... but let's see how it goes...

Comment (MarcSabatella) [inline libmscore/instrchange.cpp]: That part is easy enough to change without affecting anything else.  As mentioned, I have mixed feelings about that as well.  It maximizes "compatibility", but it's compatibility for a buggy behavior.  Removing that chunk of code would automatically fix some scores.

**Pull request merged**